twisted/internet/iocpreactor/iocpsupport/iocpsupport.pyx

```diff
@@ -214,55 +218,89 @@ def makesockaddr(object buff):
 cdef object _makesockaddr(sockaddr *addr, Py_ssize_t len):
     cdef sockaddr_in *sin
     cdef sockaddr_in6 *sin6
-    cdef char buff[256]
+    cdef WCHAR buff[256]
+    cdef DWORD buffWcharLen = <DWORD>(sizeof(buff) / sizeof(WCHAR)) - 1
     cdef int rc
-    cdef DWORD buff_size = sizeof(buff)
     if not len:
         return None
+
+    memset(buff, 0, sizeof(buff))
+
     if addr.sa_family == AF_INET:
         sin = <sockaddr_in *>addr
-        return PyBytes_FromString(inet_ntoa(sin.sin_addr)), ntohs(sin.sin_port)
+        rc = WSAAddressToStringW(addr, sizeof(sockaddr_in), NULL, buff,
+                                 &buffWcharLen)
+        if rc == SOCKET_ERROR:
+            raise_error(0, 'WSAAddressToStringW')
+        sa_port = ntohs(sin.sin_port)
+        host = PyUnicode_FromWideChar(buff, wcslen(buff))
+        host, port = host.rsplit(u':', 1)
+        port = int(port)
+        assert port == sa_port
+
+        return host, port
     elif addr.sa_family == AF_INET6:
         sin6 = <sockaddr_in6 *>addr
-        rc = WSAAddressToStringA(addr, sizeof(sockaddr_in6), NULL, buff, &buff_size)
+        rc = WSAAddressToStringW(addr, sizeof(sockaddr_in6), NULL, buff,
+                                 &buffWcharLen)
         if rc == SOCKET_ERROR:
-            raise_error(0, 'WSAAddressToString')
-        host, sa_port = PyBytes_FromString(buff), ntohs(sin6.sin6_port)
-        host, port = host.rsplit(':', 1)
+            raise_error(0, 'WSAAddressToStringW')
+        sa_port = ntohs(sin6.sin6_port)
+        host = PyUnicode_FromWideChar(buff, wcslen(buff))
+        host, port = host.rsplit(u':', 1)
         port = int(port)
         assert host[0] == '['
         assert host[-1] == ']'
         assert port == sa_port
+
         return host[1:-1], port
     else:
-        return PyBytes_FromStringAndSize(addr.sa_data, sizeof(addr.sa_data))
+        raise_error(0, "unsupported address family %d" % (addr.sa_family))
 
 
 cdef object fillinetaddr(sockaddr_in *dest, object addr):
     cdef unsigned short port
-    cdef unsigned long res
-    cdef char *hoststr
+    cdef WCHAR hostStr[256]
```

Comment:
In Python code, I'm usually against comments, but in C (or C-esque, I guess?) code it seems important to call stuff out.  256 is an appropriate value here because that's a little longer than the longest possible domain name: http://stackoverflow.com/questions/32290167/what-is-the-maximum-length-of-a-dns-name#32294443 - right?  Stack allocation always makes me nervous :).  (It looks like `hoststr` used to just be leaked though?)